cpp/src/rolling/rolling_collect_list.cu

```diff
@@ -0,0 +1,170 @@
+                                                           rmm::cuda_stream_view stream)
+{
+  // First, reduce offsets column by key, to identify the number of times
+  // an offset appears.
+  // Next, scatter the count for each offset (except the first and last),
+  // into a column of N `0`s, where N == number of child rows.
+  // For the example above:
+  //   offsets        == [0, 2, 5, 8, 11, 13]
+  //   scatter result == [0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0]
+  //
+  // If the above example had an empty list row at index 2,
+  // the same columns would look as follows:
+  //   offsets        == [0, 2, 5, 5, 8, 11, 13]
+  //   scatter result == [0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 0, 1, 0]
+  //
+  // Note: To correctly handle null list rows at the beginning of
+  // the output column, care must be taken to skip the first `0`
+  // in the offsets column, when running `reduce_by_key()`.
+  // This accounts for the `0` added by default to the offsets
+  // column, marking the beginning of the column.
+
+  auto const num_child_rows{
+    cudf::detail::get_value<size_type>(offsets, offsets.size() - 1, stream)};
+
+  rmm::device_uvector<size_type> scatter_values(offsets.size(), stream);
+  rmm::device_uvector<size_type> scatter_keys(offsets.size(), stream);
+  auto reduced_by_key =
+    thrust::reduce_by_key(rmm::exec_policy(stream),
+                          offsets.begin<size_type>() + 1,  // Skip first 0 in offsets.
+                          offsets.end<size_type>(),
+                          thrust::make_constant_iterator<size_type>(1),
+                          scatter_keys.begin(),
+                          scatter_values.begin());
+  auto scatter_values_end = reduced_by_key.second;
+  rmm::device_uvector<size_type> scatter_output(num_child_rows + 1, stream);
+  thrust::fill_n(rmm::exec_policy(stream), scatter_output.begin(), num_child_rows, 0);
+  thrust::scatter(rmm::exec_policy(stream),
+                  scatter_values.begin(),
+                  scatter_values_end,
+                  scatter_keys.begin(),
+                  scatter_output.begin());  // [0,0,1,0,0,1,...]
+
+  // Next, generate mapping with inclusive_scan() on scatter() result.
+  // For the example above:
+  //   scatter result == [0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0]
+  //   inclusive_scan == [0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4]
+  //
+  // For the case with an empty list at index 3:
+  //   scatter result == [0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 0, 1, 0]
+  //   inclusive_scan == [0, 0, 1, 1, 1, 3, 3, 3, 4, 4, 4, 5, 5]
+  auto per_row_mapping = make_fixed_width_column(
+    data_type{type_to_id<size_type>()}, num_child_rows, mask_state::UNALLOCATED, stream);
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         scatter_output.begin(),
+                         scatter_output.begin() + num_child_rows,
+                         per_row_mapping->mutable_view().template begin<size_type>());
```

Comment:
This is very elegant. (reduce_by_key to handle empty row list).

On a lighter note, here is a suggestion to avoid temporary memory.
```suggestion
  auto per_row_mapping = make_fixed_width_column(
    data_type{type_to_id<size_type>()}, num_child_rows, mask_state::UNALLOCATED, stream);
  auto per_row_mapping_begin = per_row_mapping->mutable_view().template begin<size_type>();
  thrust::fill_n(rmm::exec_policy(stream), per_row_mapping_begin, num_child_rows, 0);

  thrust::scatter_if(rmm::exec_policy(stream),
                     thrust::make_counting_iterator<size_type>(0),
                     thrust::make_counting_iterator<size_type>(0)+offsets.size()-1,
                     offsets.begin<size_type>(),
                     make_counting_transform_iterator(0, [offset=offsets.begin<size_type>()]
                     __device__(auto i) { return offset[i] != offset[i + 1]; },
                     per_row_mapping_begin);  // [0,0,1,0,0,3,...]

  // Next, generate mapping with inclusive_scan(max) on scatter() result.
  // For the example above:
  //   scatter result == [0, 0, 1, 0, 0, 2, 0, 0, 3, 0, 0, 1, 0]
  //   inclusive_scan == [0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4]
  //
  // For the case with an empty list at index 3:
  //   scatter result == [0, 0, 1, 0, 0, 3, 0, 0, 4, 0, 0, 5, 0]
  //   inclusive_scan == [0, 0, 1, 1, 1, 3, 3, 3, 4, 4, 4, 5, 5]
  thrust::inclusive_scan(rmm::exec_policy(stream),
                         per_row_mapping_begin,
                         per_row_mapping_begin  + num_child_rows,
                         per_row_mapping_begin,
                         thrust::maximum<size_type>{});
```

Reply:
This looks awesome.
So I did not write this code in this PR but simply moved it from the `.cuh` file (and fixed the memory write error).
I tried your code here and although the logic looks correct to me the `thrust::scatter_if` call results in a  `cudaErrorInvalidDeviceFunction`. I've found a solution by recoded it a bit to split up the stencil.

Reply:
Nice solution. 
I didn't know that using lambda in `pred` would solve this. 
`cudaErrorInvalidDeviceFunction` comes often for device lambda functions. (Especially with transform_iterator). 
Only known solutions are, 
1. use functor instead of device lambda. 
2. use __host__ and __device__ both
3. thrust placeholder expression (which is not possible always).